src/components/sidenav/sidenav.ts

```diff
@@ -0,0 +1,353 @@
+   */
+  public get width_() {
+    if (this.elementRef_.nativeElement) {
+      return this.elementRef_.nativeElement.offsetWidth;
+    }
+    return 0;
+  }
+
+  private transition_: boolean = false;
+  private openPromise_: Promise<void>;
+  private openPromiseResolve_: () => void;
+  private openPromiseReject_: () => void;
+  private closePromise_: Promise<void>;
+  private closePromiseResolve_: () => void;
+  private closePromiseReject_: () => void;
+}
+
+
+
+/**
+ * <md-sidenav-layout> component.
+ */
+@Component({
+  selector: 'md-sidenav-layout',
+  directives: [MdSidenav],
+  templateUrl: './components/sidenav/sidenav.html',
+  styleUrls: ['./components/sidenav/sidenav.css'],
+})
+export class MdSidenavLayout implements AfterContentInit {
+  @ContentChildren(MdSidenav) private drawers_: QueryList<MdSidenav>;
+
+  get start() { return this.start_; }
+  get end() { return this.end_; }
+
+  constructor(@Optional() @Host() private dir_: Dir) {
+    if (dir_) {
+      dir_.dirChange.subscribe(() => this.validateDrawers_());
+    }
+  }
+
+  ngAfterContentInit() {
+    // On changes, assert on consistency.
+    this.drawers_.changes.subscribe(() => this.validateDrawers_());
+    this.validateDrawers_();
+  }
+
+
+
+  private start_: MdSidenav;
+  private end_: MdSidenav;
+  private right_: MdSidenav;
+  private left_: MdSidenav;
+
+  private validateDrawers_() {
+    this.start_ = this.end_ = null;
+    if (this.drawers_.length === 0) {
+      throw new MdMissingSidenavException();
+    }
+
+    this.drawers_.toArray().forEach(drawer => {
+      if (drawer.align == 'end') {
+        if (this.end_) {
+          throw new MdDuplicatedSidenavException('end');
+        }
+        this.end_ = drawer;
+      } else {
+        if (this.start_) {
+          throw new MdDuplicatedSidenavException('start');
+        }
+        this.start_ = drawer;
+      }
+    });
+
+    this.right_ = this.left_ = null;
+    this.left_ = this.start_;
+    this.right_ = this.end_;
+
+    // Detect if we're LTR or RTL.
+    if (!this.dir_ || this.dir_.dir == 'ltr') {
+      this.left_ = this.start_;
+      this.right_ = this.end_;
+    } else {
+      this.left_ = this.end_;
+      this.right_ = this.start_;
+    }
+  }
+
+  private onBackdropClick_() {
+    if (this.start_ && this.start_.mode != 'side') {
+      this.start_.close();
+    }
+    if (this.end_ && this.end_.mode != 'side') {
+      this.end_.close();
+    }
+  }
+
+  private get showBackdrop_() {
+    return (this.start_ && this.start_.mode != 'side' && this.start_.opened)
+        || (this.end_ && this.end_.mode != 'side' && this.end_.opened);
+  }
+
+  /**
+   * Return the width of the sidenav, if it's in the proper mode and opened.
+   * @param MdSidenav
+   * @private
+   */
+  private widthFor_(sidenav: MdSidenav, mode: string): number {
+    if (sidenav && sidenav.mode == mode && sidenav.opened) {
+      return sidenav.width_;
+    }
+    return 0;
+  }
+
+  private get marginLeft_() {
```

Comment:
Same thing about these being functions. In general, I prefer functions over getters whenever there isn't also a setter.